src/app/util/af-main-common.cpp

```diff
@@ -227,7 +227,7 @@ EmberAfCbkeKeyEstablishmentSuite emberAfIsFullSmartEnergySecurityPresent(void)
     return cbkeKeyEstablishmentSuite;
 }
 
-static EmberStatus send(EmberOutgoingMessageType type, uint64_t indexOrDestination, EmberApsFrame * apsFrame,
+static EmberStatus send(EmberOutgoingMessageType type, void * exchangeContext, EmberApsFrame * apsFrame,
```

Comment:
This does not seem right to me.  It'll work OK for responses, but not for server-initiated communication (like attribute reports, events, etc, etc).

Now maybe such communication needs to set up an exchange somewhere higher up the stack, and maybe we need to rejigger the APIs to more clearly differentiate between responses and server-initiated communication....

It looks like this PR just rips out a bunch of existing functionality to do with bindings and whatnot.  We should instead design things so that this functionality works naturally; there is no obvious differences between outgoing communication to a specific node id and to a binding, except insofar as whether we're doing a response or not...

Please also note https://github.com/project-chip/connectedhomeip/issues/3584 -- it might be worth addressing that and then having that type be able to indicate whether this send is part of an existing exchange or the start of a new exchange.

Reply:
I'd like to move the binding concept to exchange layer, this is why I purged all binding related code in ember.

Ideally, messages are exchanged using an `ExchangeContext`, for both clients and servers (controllers and devices). The binding or peer node id will be hidden inside the `ExchangeContext`.

The client need to create the context manually, using given binding (either discovered by mDNS or specified via command line). Once the context is created, upper layer can use the context directly w/o knowing anything about binding or addressing.

The server is more simple. Once it receive a message from client requesting an exchange, it will automatically allocate and set up an `ExchangeContext`, and pass it to applications or protocols. The upper layer can use it directly w/o knowing anything about binding or addressing.

I'll highlight some essential part of the  PR to help understand the flow.

Reply:
> I'd like to move the binding concept to exchange layer

You can't do that; it's a data model concept....

> The binding or peer node id will be hidden inside the ExchangeContext

And exchange context is transient, with the lifetime of an exchange, right?  A binding is long-lived; it persists across power cycles, etc.

> The client need to create the context manually, using given binding (either discovered by mDNS or specified via command line)

Maybe we are talking about different meanings of "binding"?

> The server is more simple

The server can also have bindings, no?

Reply:
Thanks for the comment. IMO, binding translates application discriminator, attribute or configuration of a connection into under layer information.

Bindings are constructed by applications, and may contains following attributes:
 * peer node id or group id for multicast
 * type: unicast or multicast
 * authentication method: SPAKE or SIGMA or No encryption
 * required or preferred transports: BLE, WiFi or Thread

These kind of information will help all under layer components select proper configurations to set up contexts. It decoupled internal implementation details of under layer components from application. For example, application doesn't need to know the address of a peer node.

Also, I'd like to let the binding focus on initializing the message channel. All information inside bindings will be resolved to actual data, and stored in contexts, application don't need to use the binding after the establishment. an `ExchangeContext` will be provide to application, use it to exchange following messages.

For the server side, we can derive the context from the initiator message, so binding may not be mandatory. And if application want to persist the information of peer nodes, there will be APIs to retrieve data from contexts to reconstruct the binding representing peer node, like `GetPeerNodeId`, `IsExchangeSecure`, `GetTransportType` and so on.

In general. Use binding to initiate a exchange channel, use context to send/receive messages.

Reply:
> IMO, binding translates application discriminator, attribute or configuration of a connection into under layer information

"binding" has a specific definition in a CHIP context.  It translates an (endpoint id, cluster id) pair into a list of (group id or node id, endpoint id) pairs.  That is, it's a list of endpoints that are bound to the given endpoint for the given cluster.

https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/cam/system-model-spec/src/data_model/System-Model.adoc#3-binding-relationship and https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/cam/system-clusters/src/data_model/Binding-Cluster.adoc for some more information, though they are not very polished so far....

> and may contains following attributes

From the point of view of the data model, the "authentication method" is a lower-level implementation detail at the moment, and not captured in the data model concept of a "binding".  If it needs to be persistently stored at the data model level, we should adjust the data model bits accordingly....

Same for the transport bits.

> For example, application doesn't need to know the address of a peer node.

If by "address" you mean IP address, absolutely.  As you might note from the PRs above, the data model uses "address" to mean "Node id" in various cases.  It should probably stop doing that... 

> application don't need to use the binding after the establishment

Yes, that makes sense.

> For the server side, we can derive the context from the initiator message

The "server" can be the initiator, is the point.  At least if you mean the server side of a cluster.  Again, we might be using different meanings of the term "server" here.  :)

> In general. Use binding to initiate a exchange channel, use context to send/receive messages.

This makes perfect sense.

Again, in the low-level data model sending code in `af-main-common.cpp` it feels to me like we should support "send a message here" where "here" might be a binding or an existing exchange.  Whether we need support for "here" being a raw node id is an interesting question.  In past discussion, it seemed that a node id could be used to look up the relevant exchange context, which is why the code is structured the way it is right now.  But I would be fine with explicitly passing through some sort of exchange identifier when we know we are processing a message from an exchange (and hence responses should go in the same exchange) instead; whatever makes sense to the lower levels here.

So the way I see it, when we are done `af-main-common` message-sending code examines its arguments.  If it's provided an existing exchange, it uses that to send the message.  If it's provided a binding, it initiates a new exchange and sends on that.  Whether it also needs explicit nodeid/groupid forms (as it has now) or whether the "exchange" form would subsume those is TBD; we need to audit the uses of those forms to see what situations they are meant to cover.

Does that all make sense?